sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -4345,32 +4340,42 @@ object ApplyCharTypePadding extends Rule[LogicalPlan] {
 }
 
 /**
- * Removes all [[TempResolvedColumn]]s in the query plan. This is the last resort, in case some
- * rules in the main resolution batch miss to remove [[TempResolvedColumn]]s. We should run this
- * rule right after the main resolution batch.
+ * The rule `ResolveAggregationFunctions` in the main resolution batch creates
+ * [[TempResolvedColumn]] in filter conditions and sort expressions to hold the temporarily resolved
+ * column with `agg.child`. When filter conditions or sort expressions are resolved,
+ * `ResolveAggregationFunctions` will turn [[TempResolvedColumn]] to [[AttributeReference]] if it's
+ * inside aggregate functions or group expressions, or turn it to [[UnresolvedAttribute]] otherwise,
+ * hoping other rules can resolve it.
+ *
+ * This rule runs after the main resolution batch, and can still hit [[TempResolvedColumn]] if
+ * filter conditions or sort expressions are not resolved. When this happens, there is no point to
```

Comment:
It reads like a contradiction here. `ResolveAggregationFunctions` will turn `TempResolvedColumn` to `UnresolvedAttribute` and hope other rules can resolve it.

So why `RemoveTempResolvedColumn` can hit `TempResolvedColumn`? If it is not resolved, isn't it turned to `UnresolvedAttribute`?

Reply:
Let me make the comments clearer, `ResolveAggregationFunctions` will only replace `TempResolvedColumn` if filter conditions or sort expressions are resolved, otherwise, `TempResolvedColumn` will remain.

Reply:
So you mean if `TempResolvedColumn` is inside filter conditions or sort expressions and these expressions are resolved, `TempResolvedColumn` will be replaced with `AttributeReference`.


If the filter conditions or sort expressions are not resolved, `TempResolvedColumn` are remaining here.

Is it correct?

What `TempResolvedColumn` will be replaced with `UnresolvedAttribute`?

Reply:
```
if (filter conditions or sort expressions are resolved) {
  if (in agg func or group expr) {
    turn to AttributeReference
  } else {
    turn to UnresolvedAttribute
  }
} else {
  TempResolvedColumn remains
}
```